raft/src/main/java/org/apache/kafka/raft/internals/BatchAccumulator.java

```diff
@@ -37,6 +37,7 @@
     private final Time time;
     private final SimpleTimer lingerTimer;
     private final int lingerMs;
+    private final int minFlushSize;
```

Comment:
The main thing I'm wondering is if it makes sense to merge `minFlushSize` and `maxBatchSize` into a single configuration. I think that would simplify the implementation a bit since we could then check if `completed` is not empty to know whether to drain. It is clear that `minFlushSize` should be at least as large as `maxBatchSize`, but I'm not sure how useful it is for it to be larger. I was planning to keep `maxBatchSize` a static configuration, so I guess if we want `minFlushSize` to be configurable, then we need to allow for `minFlushSize` to be larger. What do you think?

Reply:
The way I thought about it is to still keep it configurable and keep it within the bounds of maxBatchSize. With that, the users would have the option of being able to flush more frequently based on size if it suits them and not wait for time bound flushes. We should still ensure a lower bound for this config otherwise, the fsyncs can become too frequent which can have adverse effects. I agree, setting it higher than `maxBatchSize` also might not be too useful as that might potentially lead to delayed fsyncs if in case the linger time has also been set to a higher value.

Having said that, IMO it might be useful to keep minFlushSize configurable which gives the users more knobs to control the behaviour based upon their needs(even though more knobs sometimes can mean more confusion). Do you think it makes sense?

Reply:
Yeah, I think there should be a configuration here; I am just a little unsure how it should affect the batch size. In the producer, for example, we just expose the batch size directly. We could do the same here, but I'm slightly inclined to keep the batch size decoupled from this configuration. Say for example, that we have a hard-coded 1MB max batch size. Then we could say that the effective batch size is equal to `min(1MB, maxUnflushedSize)`. A user _could_ then define `maxUnflushedSize` to be larger than 1MB and we would not have to deal with the implications of huge batches. Does that make sense?

@jsancio Any thoughts on this?

Reply:
I think it is important to agree on the current (`trunk`) behavior of `BatchAccumulator`. It is my understanding that the `BatchAccumulator` creates an in memory batch when `append` has been called enough times with enough data such that the sum of records sizes is greater than `maxBatchSize` (https://github.com/apache/kafka/pull/9756/files#diff-b7a2129b03764fbafab69c81e985d8ac6006d55f95307f0e21c70fb4750f61b5R145-R148) **OR** the first un-drained `append` is `lingerMs` old (https://github.com/apache/kafka/pull/9756/files#diff-b7a2129b03764fbafab69c81e985d8ac6006d55f95307f0e21c70fb4750f61b5R278).

I see two issues with the current implementation (`trunk`):

1. The draining thread will not call `drain`, even if `completed` is not empty, until `lingerMs` has expired (https://github.com/apache/kafka/pull/9756/files#diff-b7a2129b03764fbafab69c81e985d8ac6006d55f95307f0e21c70fb4750f61b5R254-R255).

Can this issue be address by checking if `completed` is non empty in `timeUntilDrain`?

2. The KafkaRaftClient doesn't expose a way to configure the `BatchAccumulator`'s `maxBatchSize`.

Can this issue be address by adding this configuration to `RaftConfig` and the second constructor to `KafkaRaftClient` (https://github.com/apache/kafka/pull/9756/files#diff-1da15c51e641ea46ea5c86201ab8f21cfee9e7c575102a39c7bae0d5ffd7de39R179)?

Reply:
> Say for example, that we have a hard-coded 1MB max batch size. Then we could say that the effective batch size is equal to min(1MB, maxUnflushedSize)

Hmm. I am sure on the right behaviour here. You probably already discussed this and I missed it but can you elaborate on why we want to have a minimum for the maximum batch size?

Are you thinking that we need to have a maximum batch size that is at least as large as the smallest record that the Kafka Controller can generate?

Reply:
Batching has a lot of implications in the system. The flush behavior is just one. It also impacts the segment size and the ability to do down-conversion efficiently since we have to read the whole batch into memory. The controller is designed to write small messages, so I do not think 1MB (say) would be much of a constraint. For example, we have avoided storing assignment state in a single message as we did with Zookeeper. We can reconsider it if ever needed, but I'd like to keep batches relatively small if possible.

Reply:
It sounds like we are basically on the same page here. I think we agree that there should be a configuration. As suggested in the other comment, I like `quorum.append.max.unflushed.bytes` to go along with `quorum.append.linger.ms`. Perhaps we could even use `quorum.append.max.linger.ms` to make it clear that it is the maximum amount of time that the server will hold onto a write before flushing.

In regard to the question of setting the batch size, I suggest we stick with the hard-coded value for now and set the effective value to be the minimum of the hard-coded value and `max.unflushed.bytes`. We can file a separate jira to consider whether this should be configurable separately or whether we can allow `max.unflushed.bytes` to specify the batch size precisely.

@vamossagar12 Does that sound reasonable to you?

Reply:
@hachikuji , @jsancio I agree to the points. I am slightly confused on the advantage of letting the config to be set higher than maxBatchSize(1MB) and then limiting the effective batch size to the minimum of `maxUnflushedBytes` and `maxBatchSize`. Won't it lead to confusion in terms of usage?
Infact, I have followed a similar approach in the PR in a different way. What I have done is restricting the value of this new config within a range of 0(which is wrong, need to change) to maxBatchSize - 1:


```            .define(QUORUM_FLUSH_MIN_SIZE_BYTES_CONFIG,
                ConfigDef.Type.INT,
                1024,
                between(0, (1024 * 1024) - 1),
                ConfigDef.Importance.MEDIUM,
                QUORUM_FLUSH_MIN_SIZE_BYTES_DOC);
```

I feel this way, what the config provides and what actually happens remains the same and effectively, this will be equivalent to `min(maxBatchsize, maxUnflushedBytes)`.

Let me know your thoughts plz.

Reply:
> I am slightly confused on the advantage of letting the config to be set higher than maxBatchSize(1MB) and then limiting the effective batch size to the minimum of maxUnflushedBytes and maxBatchSize. Won't it lead to confusion in terms of usage?

Not sure if this is the cause of the confusion, but just to be clear, what I'm referring to as the batch size is just a limit to the size of a `DefaultRecordBatch`. Records in Kafka are grouped together into batches with a schema roughly like this:
```
Batch => Size Offset Crc ... [Record]
```
It is useful to keep batches from getting too large because we don't have a way to index into a batch without storing the whole decompressed contents in memory. 

However, we can let `maxUnflushedBytes` be as large as we want. That just means that we will collect multiple batches (i.e. multiple `CompletedBatch` instances) as the code already does. So if `maxUnflushedBytes` is set to 4MB, then we will end up collecting 4 batches before we flush to disk. We could also require that `maxUnflushedBytes` be less than or equal to our desired max batch size, but if it's a similar level of effort to support the more general config, then that seems better.

> Also @hachikuji , wanted to understand how the newly proposed config; quorum.append.max.linger.ms would interplay with the existing quorum.append.linger.ms config. As per my understanding, the moment quorum.append.linger.ms is crossed, the flush would start. This happens even in this new implementation irrespective of hitting maxUnflushedBytes or not. Are you suggesting that we still hold onto writes until we hit the quorum.append.max.linger.ms thereby overriding quorum.append.linger.ms?

My expectation is that we flush after either of these are reached. So if the linger time is hit first, then we take whatever unflushed bytes we have even if they are smaller than `maxUnflushedBytes`. On the other hand, if `maxUnflushedBytes` is reached, then we ignore linger. I think this is consistent with the implementation I suggested here: https://github.com/apache/kafka/pull/9756#discussion_r552268919. Basically we override `timeUntilDrain` so that it returns 0 when `maxUnflushedBytes` is reached.

Hope that helps!